src/Symfony/Component/Translation/Bridge/Lokalise/Tests/LokaliseProviderTest.php

```diff
@@ -0,0 +1,584 @@
+                            ],
+                        ],
+                    ],
+                ],
+            ]);
+
+            $this->assertSame('PUT', $method);
+            $this->assertSame($expectedBody, $options['body']);
+
+            return new MockResponse();
+        };
+
+        $provider = $this->createProvider((new MockHttpClient([
+            $getLanguagesResponse,
+            $createLanguagesResponse,
+            $getKeysIdsForMessagesDomainResponse,
+            $getKeysIdsForValidatorsDomainResponse,
+            $createKeysForMessagesDomainResponse,
+            $createKeysForValidatorsDomainResponse,
+            $updateTranslationsResponse,
+        ]))->withOptions([
+            'base_uri' => 'https://api.lokalise.com/api2/projects/PROJECT_ID/',
+            'headers' => ['X-Api-Token' => 'API_KEY'],
+        ]), $this->getLoader(), $this->getLogger(), $this->getDefaultLocale(), 'api.lokalise.com/api2/');
+
+        $translatorBag = new TranslatorBag();
+        $translatorBag->addCatalogue(new MessageCatalogue('en', [
+            'messages' => ['a' => 'trans_en_a'],
+            'validators' => ['post.num_comments' => '{count, plural, one {# comment} other {# comments}}'],
+        ]));
+        $translatorBag->addCatalogue(new MessageCatalogue('fr', [
+            'messages' => ['a' => 'trans_fr_a'],
+            'validators' => ['post.num_comments' => '{count, plural, one {# commentaire} other {# commentaires}}'],
+        ]));
+
+        $provider->write($translatorBag);
+    }
+
+    /**
+     * @dataProvider getResponsesForOneLocaleAndOneDomain
+     */
+    public function testReadForOneLocaleAndOneDomain(string $locale, string $domain, string $responseContent, TranslatorBag $expectedTranslatorBag)
+    {
+        $response = function (string $method, string $url, array $options = []) use ($locale, $domain, $responseContent): ResponseInterface {
+            $expectedBody = json_encode([
+                'format' => 'symfony_xliff',
+                'original_filenames' => true,
+                'directory_prefix' => '%LANG_ISO%',
+                'filter_langs' => [$locale],
+                'filter_filenames' => [$domain.'.xliff'],
+            ]);
+
+            $this->assertSame('POST', $method);
+            $this->assertSame('https://api.lokalise.com/api2/projects/PROJECT_ID/files/export', $url);
+            $this->assertSame($expectedBody, $options['body']);
+
+            return new MockResponse(json_encode([
+                'files' => [
+                    $locale => [
+                        $domain.'.xliff' => [
+                            'content' => $responseContent,
+                        ],
+                    ],
+                ],
+            ]));
+        };
+
+        $loader = $this->getLoader();
+        $loader->expects($this->once())
+            ->method('load')
+            ->willReturn((new XliffFileLoader())->load($responseContent, $locale, $domain));
```

Comment:
I'm not happy with this, but I don't find the correct way to either test it, or mock it. cc @nicolas-grekas